Sources/MongoSwift/MongoCursor.swift

```diff
@@ -1,19 +1,53 @@
 import mongoc
 
+internal let ClosedCursorError = UserError.logicError(message: "Cannot advance a completed or failed cursor.")
+
 /// A MongoDB cursor.
 public class MongoCursor<T: Codable>: Sequence, IteratorProtocol {
-    /// Pointer to underlying `mongoc_cursor_t`.
-    internal let _cursor: OpaquePointer
-    /// We store these three objects to ensure that they remain in scope for as long as this cursor does.
-    private let _client: MongoClient
-    private let _connection: Connection
-    private let _session: ClientSession?
+    /// Enum for tracking the state of a cursor.
+    internal enum State {
+        /// Indicates that the cursor is still open. Stores a pointer to the `mongoc_cursor_t`, along with the source
+        /// connection, client, and possibly session to ensure they are kept alive as long as the cursor is.
+        case open(cursor: OpaquePointer, connection: Connection, client: MongoClient, session: ClientSession?)
+        case closed
+    }
+
+    /// The state of this cursor.
+    internal private(set) var state: State
+
+    /// The error that occurred while iterating this cursor, if one exists. This should be used to check for errors
+    /// after `next()` returns `nil`.
+    public private(set) var error: Error?
 
-    private var swiftError: Error?
+    /// Indicates whether this is a tailable cursor.
+    private let cursorType: CursorType
 
     /// Decoder from the `MongoCollection` or `MongoDatabase` that created this cursor.
     internal let decoder: BSONDecoder
 
+    /**
+     * Indicates whether this cursor has the potential to return more data. This property is mainly useful for
+     * tailable cursors, where the cursor may be empty but contain more results later on. For non-tailable cursors,
+     * the cursor will always be dead as soon as `next()` returns `nil`, or as soon as `nextOrError()` returns `nil` or
+     * throws an error.
+     */
+    public var isAlive: Bool {
+        if case .open = self.state {
+            return true
+        }
+        return false
+    }
+
+    /// Returns the ID used by the server to track the cursor. `nil` until mongoc actually talks to the server by
+    /// iterating the cursor, and `nil` after mongoc has fetched all the results from the server.
+    internal var id: Int64? {
+        guard case let .open(cursor, _, _, _) = self.state else {
+            return nil
+        }
+        let id = mongoc_cursor_get_id(cursor)
+        return id == 0 ? nil : id
```

Comment:
nitpick: I think there's a meaningful difference between 0 and `nil` in this case. 0 means the cursor is closed server side, but nil might indicate the server was never actually initialized or even run.

Reply:
we are kind of limited here by the API mongoc provides -- a `0` return value from `mongoc_cursor_get_id` can indicate either of those cases:
> This number is zero until the driver actually uses a server when executing the query, and after it has fetched all results from the server.

we could consider adding some kind of boolean flag in `MongoCursor` that tracks if we've iterated at all, and if we haven't, return `nil` here, and otherwise directly return the 0. do you think that's worth it?

Reply:
nah, let's just leave this as-is, and consider it in a future where we're not using libmongoc